lightning/src/chain/chainmonitor.rs

```diff
@@ -32,45 +32,76 @@ use chain::chaininterface::{BroadcasterInterface, FeeEstimator};
 use chain::channelmonitor::{ChannelMonitor, ChannelMonitorUpdate, Balance, MonitorEvent, TransactionOutputs};
 use chain::transaction::{OutPoint, TransactionData};
 use chain::keysinterface::Sign;
+use util::atomic_counter::AtomicCounter;
 use util::logger::Logger;
 use util::events;
 use util::events::EventHandler;
 use ln::channelmanager::ChannelDetails;
 
 use prelude::*;
-use sync::{RwLock, RwLockReadGuard};
+use sync::{RwLock, RwLockReadGuard, Mutex};
 use core::ops::Deref;
+use core::sync::atomic::{AtomicBool, Ordering};
+
+#[derive(Clone, Copy, Hash, PartialEq, Eq)]
+/// A specific update stored in a `MonitorUpdateId`, separated out to make the contents entirely
+/// opaque.
+pub(crate) enum MonitorUpdate {
+	/// An update which was generated by the [`ChannelManager`] (via our `chain::Watch`
+	/// implementation). This corresponds to an actual [`ChannelMonitorUpdate::update_id`] field
+	/// and [`ChannelMonitor::get_latest_update_id`].
+	MonitorUpdateId(u64),
+	/// An update which was generated during blockchain processing. The ID here is specific to the
+	/// generating [`ChainMonitor`] and does *not* correspond to any on-disk IDs.
+	SyncPersistId(u64),
+}
+
+/// An opaque identifier describing a specific [`Persist`] method call.
+#[derive(Clone, Copy, Hash, PartialEq, Eq)]
+pub struct MonitorUpdateId {
+	pub(crate) contents: MonitorUpdate,
```

Comment:
I think the struct name `MonitorUpdate` is a little conflicting/confusing with `ChannelMonitorUpdate`. Also would be nice to avoid confusing destructuring like `MonitorUpdateId { MonitorUpdate::MonitorUpdateId ... }`

Maybe instead:

```
pub struct MonitorUpdateId {
    pub(crate) inner: u64,
    pub(crate) origin: UpdateOrigin
}

pub(crate) enum UpdateOrigin {
    ChainSync,
    Offchain,
}
```